tests/CppWinRTComServerTests.cpp

```diff
@@ -0,0 +1,137 @@
+#include "pch.h"
+
+#include "common.h"
+#undef GetCurrentTime
+#define WINRT_CUSTOM_MODULE_LOCK
+#include <wil/cppwinrt_notifiable_module_lock.h>
+#include <winrt/Windows.Foundation.h>
+#include <wil/cppwinrt_register_com_server.h>
+#include <wil/resource.h>
+#include <string_view>
+#include <thread>
+#include <chrono>
+
+using namespace std::string_view_literals;
+
+wil::unique_event _comExit;
+
+void notifier()
+{
+    _comExit.SetEvent();
+}
+
+struct MyServer : winrt::implements<MyServer, winrt::Windows::Foundation::IStringable>
+{
+    winrt::hstring ToString()
+    {
+        return L"MyServer from Server";
+    }
+};
+
+struct BuggyServer : winrt::implements<BuggyServer, winrt::Windows::Foundation::IStringable>
+{
+    BuggyServer()
+    {
+        throw winrt::hresult_access_denied{};
+    }
+    winrt::hstring ToString()
+    {
+        return L"BuggyServer from Server";
+    }
+};
+
+auto create_my_server_instance()
+{
+    return winrt::create_instance<winrt::Windows::Foundation::IStringable>(winrt::guid_of<MyServer>(), CLSCTX_LOCAL_SERVER);
+}
+
+TEST_CASE("CppWinRTComServerTests::DefaultNotifiableModuleLock", "[cppwinrt_com_server]")
+{
+    _comExit.create();
+
+    wil::notifiable_module_lock::set_notifier(notifier);
+
+    winrt::init_apartment();
+
+    {
+        auto server{winrt::make<MyServer>()};
+        REQUIRE(winrt::get_module_lock() == 1);
+    }
+
+    _comExit.wait();
+
+    REQUIRE(!winrt::get_module_lock());
+}
+
+TEST_CASE("CppWinRTComServerTests::RegisterComServer", "[cppwinrt_com_server]")
+{
+    winrt::init_apartment();
+
+    {
+        auto revoker = wil::register_com_server<MyServer>();
+        auto instance = create_my_server_instance();
+        REQUIRE(winrt::get_module_lock() == 1);
+    }
+    REQUIRE(!winrt::get_module_lock());
+    try
+    {
+        auto instance = create_my_server_instance();
+    }
+    catch (winrt::hresult_error const& e)
+    {
+        REQUIRE(e.code() == REGDB_E_CLASSNOTREG);
+    }
+}
+
+TEST_CASE("CppWinRTComServerTests::RegisterComServerThrowIsSafe", "[cppwinrt_com_server]")
+{
+    winrt::init_apartment();
+
+    {
+        auto revoker = wil::register_com_server<BuggyServer>();
+        try
+        {
+            auto instance = winrt::create_instance<winrt::Windows::Foundation::IStringable>(winrt::guid_of<BuggyServer>(), CLSCTX_LOCAL_SERVER);
+            REQUIRE(false);
+        }
+        catch (winrt::hresult_error const& e)
+        {
+            REQUIRE(e.code() == E_ACCESSDENIED);
+        }
+    }
+}
+
+winrt::Windows::Foundation::IAsyncAction create_instance_after_5_s()
+{
+    using namespace std::chrono_literals;
+    co_await winrt::resume_after(5s);

```

Comment:
What benefit does waiting here have? 5 seconds is quite a while to delay test execution for. From what I can gather, there's not really much benefit over doing something simple like `co_await winrt::resume_background();`

Reply:
Just simulating a real world usage where the server is created for a while before an instance is created. I guess what you are saying is this is not something worth testing for? I don't know whether there had been edge cases where bugs only manifest in this case (as opposed to the case where the executable is spawned per activation request), so it might indeed be redundant.

Reply:
In this case, it seems like you're more testing order of operations:
1. Register server
2. Bump module ref count w/ coroutine
3. Bump module ref count with `CoCreateInstance`
4. Drop both module ref counts
5. Main thread completes operation. 

The only real synchronization point is between 4 & 5, but that's already being handled by your event wait. The sleep doesn't _really_ help test much more than to effectively ensure that the wait for 5 happens before 3. You can get the same effect with additional synchronization. E.g. something maybe like:
```c++
TEST_CASE("CppWinRTComServerTests::NotifierAndRegistration", "[cppwinrt_com_server]")
{
    wil::unique_event moduleEvent(wil::EventOptions::None);
    wil::unique_event coroutineRunning(wil::EventOptions::None);
    wil::unique_event coroutineContinue(wil::EventOptions::None);

    wil::notifiable_module_lock::set_notifier([&] {
        moduleEvent.SetEvent();
    });

    winrt::init_apartment();

    auto revoker = wil::register_com_server<MyServer>();

    [&]() {
        co_await winrt::resume_background();
        coroutineRunning.SetEvent();

        coroutineContinue.wait();
        auto instance = create_my_server_instance();
        REQUIRE(winrt::get_module_lock() == 2);
    }();

    coroutineRunning.wait();
    REQUIRE(winrt::get_module_lock() == 1); // Coroutine bumped count

    coroutineContinue.SetEvent();
    moduleEvent.wait();

    REQUIRE(winrt::get_module_lock() == 0);
}
```

Reply:
```
coroutineContinue.SetEvent();
moduleEvent.wait();
```
Wouldn't it technically means `moduleEvent.wait()` can execute after whatever is after `coroutineContinue.wait()`?

Reply:
That's going to be true, even with the sleep (just unlikely). But that's also not entirely bad. You could also technically validate that the even is not set before calling `coroutineContinue.SetEvent()`